src/ProductManifold.jl

```diff
@@ -0,0 +1,478 @@
+@doc doc"""
+    ProductManifold{TM<:Tuple, TRanges<:Tuple, TSizes<:Tuple} <: Manifold
+
+Product manifold $M_1 \times M_2 \times \dots \times M_n$ with product geometry.
+`TRanges` and `TSizes` statically define the relationship between representation
+of the product manifold and representations of point, tangent vectors
+and cotangent vectors of respective manifolds.
+
+# Constructor
+
+    ProductManifold(M_1, M_2, ..., M_n)
+
+generates the product manifold $M_1 \times M_2 \times \dots \times M_n$.
+Alternatively, the same manifold can be contructed using the `×` operator:
+`M_1 × M_2 × M_3`.
+"""
+struct ProductManifold{TM<:Tuple} <: Manifold
+    manifolds::TM
+end
+
+ProductManifold(manifolds::Manifold...) = ProductManifold{typeof(manifolds)}(manifolds)
+
+function cross(M1::Manifold, M2::Manifold)
+    return ProductManifold(M1, M2)
+end
+
+function cross(M1::ProductManifold, M2::Manifold)
+    return ProductManifold(M1.manifolds..., M2)
+end
+
+function cross(M1::Manifold, M2::ProductManifold)
+    return ProductManifold(M1, M2.manifolds...)
+end
+
+function cross(M1::ProductManifold, M2::ProductManifold)
+    return ProductManifold(M1.manifolds..., M2.manifolds...)
+end
+
+"""
+    submanifold(M::ProductManifold, i::Integer)
+
+Extract the `i`th factor of the product manifold `M`.
+"""
+function submanifold(M::ProductManifold, i::Integer)
+    return M.manifolds[i]
+end
+
+"""
+    submanifold(M::ProductManifold, i::Val)
+
+Extract the factor of the product manifold `M` indicated by indices in `i`.
+For example, for `i` equal to `Val((1, 3))` the product manifold constructed
+from the first and the third factor is returned.
+"""
+function submanifold(M::ProductManifold, i::Val)
```

Comment:
Would it also be good to have a generic `submanifold(M::ProductManifold, i) = submanifold(M, Val((i...))` so users can get the type-stability but also use convenient `submanifold(M, 1:3)`?